File: src/main/java/com/grosner/dbflow/sql/migration/UpdateTableMigration.java
package com.grosner.dbflow.sql.migration;

import android.database.sqlite.SQLiteDatabase;

import com.grosner.dbflow.sql.Query;
import com.grosner.dbflow.sql.Update;
import com.grosner.dbflow.sql.builder.Condition;
import com.grosner.dbflow.sql.builder.ConditionQueryBuilder;
import com.grosner.dbflow.sql.builder.QueryBuilder;
import com.grosner.dbflow.structure.Model;

/**
 * Author: andrewgrosner
 * Contributors: { }
 * Description: Provides a simple way to update a table's field or fields quickly in a migration. It ties an SQLite {@link com.grosner.dbflow.sql.Update}
 * to migrations whenever we want to batch update tables in a structured manner.
 */
public class UpdateTableMigration<ModelClass extends Model> extends BaseMigration implements Query {

    /**
     * The query to use
     */
    private QueryBuilder mQuery;

    /**
     * Builds the conditions for the WHERE part of our query
     */
    private ConditionQueryBuilder<ModelClass> mWhereConditionQueryBuilder;

    /**
     * The table to update
     */
    private final Class<ModelClass> mTable;

    /**
     * The conditions to use to set fields in the update query
     */
    private ConditionQueryBuilder<ModelClass> mSetConditionQueryBuilder;

    /**
     * Creates an update migration.
     * @param table The table to update
     * @param migrationVersion The version of the db to update at.
     */
    public UpdateTableMigration(Class<ModelClass> table, int migrationVersion) {
        super(migrationVersion);
        mTable = table;
    }

    /**
     * This will append a SET columnName = value to this migration. It will execute each of these in succession with the order
     * that this is called.
     *
     * @param conditions The condition to append
     * @return
     */
    public UpdateTableMigration<ModelClass> set(Condition... conditions) {
        if (mSetConditionQueryBuilder == null) {
            mSetConditionQueryBuilder = new ConditionQueryBuilder<ModelClass>(mTable);
        }

        mSetConditionQueryBuilder.putConditions(conditions);
        return this;
    }

    public UpdateTableMigration<ModelClass> where(Condition...conditions) {
        if (mWhereConditionQueryBuilder == null) {
            mWhereConditionQueryBuilder = new ConditionQueryBuilder<ModelClass>(mTable);
        }

        mWhereConditionQueryBuilder.putConditions(conditions);
        return this;
    }

    @SuppressWarnings("unchecked")
    @Override
    public void onPreMigrate() {
        mQuery = new QueryBuilder(new Update().table(mTable).set(mSetConditionQueryBuilder).where(mWhereConditionQueryBuilder).getQuery());
    }

    @Override
    public void migrate(SQLiteDatabase database) {
        database.execSQL(mQuery.getQuery());
    }

    @Override
    public void onPostMigrate() {
        // make fields eligible for GC
        mQuery = null;
        mSetConditionQueryBuilder = null;
    }

    @Override
    public String getQuery() {
        return mQuery.getQuery();
    }
}

File: dbflow-common/src/main/kotlin/com/raizlabs/dbflow5/migration/UpdateTableMigration.kt
package com.raizlabs.dbflow5.migration

import kotlin.reflect.KClass
import com.raizlabs.dbflow5.database.DatabaseWrapper
import com.raizlabs.dbflow5.query.BaseQueriable
import com.raizlabs.dbflow5.query.OperatorGroup
import com.raizlabs.dbflow5.query.SQLOperator
import com.raizlabs.dbflow5.query.update

/**
 * Description: Provides a simple way to update a table's field or fields quickly in a migration.
 * It ties an SQLite [com.raizlabs.android.dbflow.sql.language.Update]
 * to migrations whenever we want to batch update tables in a structured manner.
 */
open class UpdateTableMigration<T : Any>
/**
 * Creates an update migration.
 *
 * @param table The table to update
 */
(
    /**
     * The table to update
     */
    private val table: KClass<T>) : BaseMigration() {

    /**
     * Builds the conditions for the WHERE part of our query
     */
    private val whereOperatorGroup: OperatorGroup by lazy { OperatorGroup.nonGroupingClause() }

    /**
     * The conditions to use to set fields in the update query
     */
    private val setOperatorGroup: OperatorGroup by lazy { OperatorGroup.nonGroupingClause() }

    val updateStatement: BaseQueriable<T>
        get() = update(table)
            .set(setOperatorGroup)
            .where(whereOperatorGroup)

    /**
     * This will append a condition to this migration. It will execute each of these in succession with the order
     * that this is called.
     *
     * @param conditions The conditions to append
     */
    fun set(vararg conditions: SQLOperator) = apply {
        setOperatorGroup.andAll(*conditions)
    }

    fun where(vararg conditions: SQLOperator) = apply {
        whereOperatorGroup.andAll(*conditions)
    }

    override fun migrate(database: DatabaseWrapper) {
        updateStatement.execute(database)
    }
}
